### PR TITLE
refactor: use regions in `SearchEnv`

### LINKED ISSUE #3896: Refactor Datalog solver to use regions
We should refactor the Datalog engine to use regions (while collecting information about the refactoring for the case study).

I propose the following approach:

- [ ] Declare a new region in `solveWithStratification` (I think that's the entry point, right?)
- [ ] Add a `r: Region[r]` parameter to every function in `Fixpoint` which appears to need such an argument. (Need to discuss if it should be first or last.)
- [ ] Refactor the Arrays in `Fixpoint.Ram` to use regions.
- [ ] Refactor the Arrays in `Fixpoint.Ast` to use regions (may require tweaks to `Lowering`.

Each should be a separate PR. Possibly with sub-PRs, if feasible.

While doing this work, keep notes of interesting observations. E.g. is the refactoring straightforward? Are there interesting choices to be made? Are some restructuring required or not?

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes #3896.

With this PR the debugging messages will now be erased by the optimizer. This is probably just the way it is for now.

The remaining impure methods are related to debug messages and the use of `Global.newId`.

The remaining unsafe Pure casts are from the above and then the unsafe implementation of `Tuple`

### CODE REVIEW

File: main/src/library/Fixpoint/Interpreter.flix [hunk 1]
```diff
@@ -76,20 +76,20 @@ namespace Fixpoint {
...
+    type alias SearchEnv[v: Type, r: Region] = (Array[Tuple[v], r], Array[v, r])
 
-    def interpret(r: Region[r], stmt: RamStmt[v]): Database[v, r] & Impure with Order[v], ToString[v] =
+    def interpret(r: Region[r], stmt: RamStmt[v]): Database[v, r] & r with Order[v], ToString[v] =
         interpretWithDatabase(new MutMap(r), stmt)
 
-    def interpretWithDatabase(db: Database[v, r], stmt: RamStmt[v]): Database[v, r] & Impure with Order[v], ToString[v] =
-        notifyPreInterpret(stmt);
+    def interpretWithDatabase(db: Database[v, r], stmt: RamStmt[v]): Database[v, r] & r with Order[v], ToString[v] =
+        let _ = notifyPreInterpret(stmt) as & Pure;
```
Comment (magnus-madsen) [inline main/src/library/Fixpoint/Interpreter.flix:85]: If the cast is not on the call to notifyPre... but on the whole body of the function then it should not be removed, right?

Comment (magnus-madsen) [reply main/src/library/Fixpoint/Interpreter.flix:85]: But then of course that cast has to be to r.

**Comment by JonathanStarup** [comment]
@magnus-madsen fixed

**Comment by magnus-madsen** [comment]
I notice that you use `r` as an effect, but you should use `Read(r)` and/or `Write(r)` since thats the future.

**Comment by JonathanStarup** [comment]
> I notice that you use `r` as an effect, but you should use `Read(r)` and/or `Write(r)` since thats the future.

I've changed it when regions are used but not for Impure. Is this consistent with other places, or should & be entirely removed?

**Pull request merged**